lib/internal/url.js

```diff
@@ -730,18 +754,25 @@ class URL {
     return this[context].search;
   }
 
-  set search(search) {
+  set search(value) {
     if (!isURLThis(this))
       throw new ERR_INVALID_THIS('URL');
-    search = toUSVString(search);
-    updateUrl(this[context].href, updateActions.kSearch, search, this.#onParseComplete);
-    initSearchParams(this[searchParams], this[context].search);
+    updateUrl(this[context].href, updateActions.kSearch, toUSVString(value), this.#onSearchUpdate);
   }
 
   // readonly
   get searchParams() {
     if (!isURLThis(this))
       throw new ERR_INVALID_THIS('URL');
+    // Create URLSearchParams on demand to greatly improve the URL performance.
+    if (this[searchParams] == null) {
+      this[searchParams] = new URLSearchParams(this[context].search);
+      this[searchParams][context] = this;
+    } else if (this[searchParams][kDirty]) {
```

Comment:
You need to check the dirty bit a lot more often than just here. `URL` and `URLSearchParams`. Consider this:

```js
const u = new URL('http://abc.com/?q=old');
const s = u.searchParams;
u.href = 'http://abc.com/?q=new';
console.log(s.get('q'));
```

It should print `new` but I suspect it prints `old`.

Or consider this:

```js
const u = new URL('http://abc.com/?key1=value1&key2=value2');
const s = u.searchParams;
for (const k of s.keys()) {
  console.log(k); // this should print key1, then key3
  u.href = 'http://abc.com/?key2=value2&key3=value3';
}
```

Ultimately, it's quite complicated to get the dirty bit respected everywhere. I suggest doing away with the dirty bit, and just re-parse parameters whenever 1) there has been a change and 2) `this[searchParams]` is populated. It's simpler, less likely to break, and if folks have accessed `searchParams` in the past they probably wouldn't mind the performance hit.

(Also, I suggest adding these two test cases to WPT if they are not there already.)

Reply:
In case you are wondering where in the spec the second test case above is specified, it's [here](https://webidl.spec.whatwg.org/#ref-for-dfn-iterator-prototype-object%E2%91%A3). Notice that on every `next` call, it recomputes the "[value pairs to iterate over](https://webidl.spec.whatwg.org/#dfn-value-pairs-to-iterate-over)", which for [URLSearchParam](https://url.spec.whatwg.org/#ref-for-dfn-value-pairs-to-iterate-over) is the current list of key–value pairs.

Reply:
This is a nice catch. Thank you for your review. I expected this to be in the web platform tests. Only creating URLSearchParams on `searchParams` getter, and auto-updating `searchParams` on `onParseComplete` serves the same goal of this pull request, and it's much more cleaner. Thank you for the suggestion!